tools/reduce.py

```diff
@@ -0,0 +1,303 @@
+#!/usr/bin/env python3
+
+import argparse
+import contextlib
+import copy
+from dataclasses import dataclass
+import json
+import logging
+import os
+import pathlib
+import re
+import shlex
+import shutil
+import stat
+import subprocess
+import sys
+import tempfile
+import textwrap
+from typing import List
+
+# https://clang.llvm.org/docs/ClangCommandLineReference.html
+# fmt: off
+PATH_FLAGS = [
+    "-fcrash-diagnostics-dir",
+    "-B",
+    "-F",
+    "-I" "/I", "--include-directory",
+    "--amdgpu-arch-tool",
+    "--cuda-path",
+    "--cxx-isystem",
+    "-fbuild-session-file",
+    "-fmodule-file", # ignoring -fmodule-file=name=x
+    "-fmodules-cache-path",
+    "-fmodules-user-build-path",
+    "-fprebuilt-module-path",
+    "--hip-path",
+    "-idirafter",
+    "-idirafter", "--include-directory-after",
+    "-iframework,"
+    "-iframeworkwithsysroot",
+    "-imacros", "--imacros",
+    "-include", "--include",
+    "-include-pch",
+    "-iprefix", "--include-prefix",
+    "-iquote",
+    "-isysroot",
+    "-isystem",
+    "-system-after",
+    "-ivfsoverlay",
+    "-iwithprefix", "--include-with-prefix", "--include-with-prefix-after",
+    "-iwithprefixbefore", "--include-with-prefix-before",
+    "-iwithsysroot",
+    "--libomptarget-amdgpu-bc-path", "--libomptarget-amdgcn-bc-path",
+    "--libomptarget-nvptx-bc-path",
+    "--nvptx-arch-tool",
+    "--ptxas-path",
+    "--rocm-path",
+    "-stdlib\+\+-isystem",
+    "--system-header-prefix", "--no-system-header-prefix"
+]
+# fmt: on
+
+
+class CompdbEntry:
+    filepath: pathlib.Path
+    directory: pathlib.Path
+    arguments: List[str]
+
+    def __init__(self, entry):
+        self.filepath = pathlib.Path(entry["file"])
+        self.directory = pathlib.Path(entry["directory"])
+        try:
+            self.arguments = shlex.split(entry["command"])
+        except KeyError:
+            self.arguments = entry(["arguments"])
+
+    def change_tu_filepath(self, new_path: str):
+        old_path = str(self.filepath)
+        for i, arg in enumerate(self.arguments):
+            if old_path in arg:
```

Comment:
Tell me about these paths... can these be empty, or short like . or a relative path? If so this could get wild.

Reply:
The path points to an implementation file as per the `file` key in https://clang.llvm.org/docs/JSONCompilationDatabase.html

It could be relative or absolute. It can't be empty or `.`.

Reply:
I'm imagining a situation like you have a path `foo.c` and a path `bar/foo.c` and, similar with the arguments which could be prefixes, the replacements are surprising.

If we understood the format of the arguments a bit better, like `--foo=bar` we only look for files in the *bar* part and matched whole strings, could be more reliable.

Realize that this is just a support tool and it doesn't need to be *super* robust. Another way to approach it would be to explain the "happy path" common use case and provide a bit more context around that, and then some pointers if there's common failure modes like you need to add a flag or something, and leave the rest to future users to elaborate.

Reply:
I added a comment describing why I think that failure mode is highly unlikely in this situation. Could you take a second look to see if that's satisfactory?